pages/Product/components/Content/spec.jsx

```diff
@@ -1,47 +1,354 @@
+  const reviewsProps = Reviews.props();
+  expect(reviewsProps).toMatchObject({ productId });
+  expect(reviewsProps.variantId).toBeUndefined();
+};
+
+describe('<ProductContent />', () => {
+  let mockedState;
 
   beforeEach(() => {
-    jest.resetModules();
+    mockedState = createState(mockedProductState);
   });
 
-  it('should render', () => {
-    const { productId } = basicProductState.product.currentProduct;
-    defaultContext.productId = productId;
-    const cmp = createComponent(basicProductState);
-    expect(cmp).toMatchSnapshot();
-    expect(cmp.find('ProductHeader').length).toEqual(1);
-    expect(cmp.find('Characteristics').length).toEqual(1);
-    expect(cmp.find('Options').length).toEqual(1);
-    expect(cmp.find('TaxDisclaimer').length).toEqual(1);
+  describe('rendering when all product data is available', () => {
+    it('should render a product without variants as expected', () => {
+      const productId = 'product_5';
+      const variantId = null;
+
+      const { component } = createComponent(mockedState, { productId });
+      runComponentTests(component, { productId, variantId });
+    });
+
+    it('should render a variant product as expected', () => {
+      const productId = 'product_1';
+      const variantId = 'product_2';
+
+      const { component } = createComponent(mockedState, { productId: variantId });
+      runComponentTests(component, { productId, variantId });
+    });
+
+    it('should render variant product updates as expected', () => {
+      const productId = 'product_1';
+      const variantId = 'product_2';
+      const variantIdUpdate = 'product_3';
+
+      // Render the component with a base product.
+      const { component } = createComponent(mockedState, { productId: variantId });
+      runComponentTests(component, { productId, variantId });
+
+      // Update the productId with the id of another variant.
+      component.setProps({ productId: variantIdUpdate });
+      runComponentTests(component, { productId, variantId: variantIdUpdate });
+    });
+  });
+
+  describe('rendering when product data is not available yet', () => {
+    describe('simple product', () => {
+      const productId = 'product_5';
+      const variantId = null;
+
+      let component;
+      let store;
+
+      beforeAll(() => {
+        // Prepare state with an unavailable base product.
+        delete mockedState.product.productsById[productId];
+        ({ component, store } = createComponent(mockedState, { productId }));
+      });
+
+      it('should initialize as expected', () => {
+        runComponentTests(component, { productId, variantId });
+      });
+
+      it('should render as expected when data became available', () => {
+        // Update the store with complete product data.
+        store.dispatch(updateState(mockedProductState));
+        component.update();
+
+        runComponentTests(component, { productId, variantId });
+      });
+    });
+
+    describe('product with variants', () => {
```

Comment:
I've tried to improve the `describe` and `it` blocks. Additionally i removed some comments, since their wrapping block should now describe what's going on inside.

In general i also prefer to test a component without its connector. But this one is special. It's usually only parametrized with a `productId` prop. By utilizing its selectors it determines two other props and generates a `productId` and a `variantId`. So this component will be never be used without its connector.

Using an unconnected component would make the test more complex and increase the danger that future code changes affect the component without noticing.

So i would be happy to keep the mocked store like it is for this test.